components/org.wso2.identity.event.websubhub.publisher/src/main/java/org/wso2/identity/event/websubhub/publisher/service/WebSubEventSubscriberImpl.java

```diff
@@ -0,0 +1,195 @@
+            return false;
+        }
+    }
+
+    @Override
+    public boolean unsubscribe(List<String> topics, String callbackUrl, String tenantDomain)
+            throws WebhookMgtException {
+
+        try {
+            // Create a list to hold all the futures
+            List<CompletableFuture<Boolean>> futures = new ArrayList<>();
+
+            for (String topic : topics) {
+                futures.add(CompletableFuture.supplyAsync(() -> {
+                    try {
+                        makeSubscriptionAPICall(constructHubTopic(topic, tenantDomain),
+                                getWebSubBaseURL(),
+                                WebSubHubAdapterConstants.Http.UNSUBSCRIBE,
+                                callbackUrl);
+                        log.debug("WebSub Hub unsubscription successful for topic: " + topic +
+                                " with callback URL: " + callbackUrl + " in tenant: " + tenantDomain);
+                        return true;
+                    } catch (WebSubAdapterException e) {
+                        log.error("Error unsubscribing to topic: " + topic, e);
+                        return false;
+                    }
+                }));
+            }
+
+            // Combine all the futures into a single future and wait for completion
+            CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+            // Wait for all futures to complete
+            allFutures.join();
+
+            // Check if any unsubscription failed
+            return futures.stream().allMatch(CompletableFuture::join);
+
+        } catch (Exception e) {
+            log.error("Unexpected error during unsubscription process", e);
+            return false;
+        }
+    }
+
+    private void makeSubscriptionAPICall(String topic, String webSubHubBaseUrl, String operation, String callbackUrl)
+            throws WebSubAdapterException {
+
+        String subscriptionUrl = buildSubscriptionURL(topic, webSubHubBaseUrl, operation, callbackUrl);
+
+        ClientManager clientManager = WebSubHubAdapterDataHolder.getInstance().getClientManager();
+        HttpPost httpPost = clientManager.createHttpPost(subscriptionUrl, null);
+
+        WebSubHubCorrelationLogUtils.triggerCorrelationLogForRequest(httpPost);
+        final long requestStartTime = System.currentTimeMillis();
+
+        CompletableFuture<HttpResponse> future = clientManager.executeAsync(httpPost);
+
+        future.thenAccept(response -> {
+            try {
+                handleSubscriptionResponse((CloseableHttpResponse) response, httpPost, topic, operation,
+                        requestStartTime);
+            } catch (IOException | WebSubAdapterException e) {
+                log.error("Handling subscription response failed. ", e);
+            }
+        }).exceptionally(ex -> {
+            log.error("Subscription API call failed. ", ex);
+            return null;
+        });
```

Comment:
Having this segment means this always wait for the response right.
So overall it looks to me each call is invoking in a synchronous manner